### ISSUE TITLE
Documentation (|||) not allowed for type constructors after the public export

### ISSUE THREAD

**Issue by FranckS** [idris-lang/Idris2#468]
Hi there,

The places where documentation (which uses |||) is accepted is currently a bit weird and inconsistent in respect to "public export" annotations.
Documentation before "public export" for functions is accepted BEFORE the "public export" (but ignored, i.e. :doc won't produce it), *and* AFTER it (and works there).
However, documentation for type constructor is accepted only before the "public export" (and works there), but not afterwards.
So in order to fully document code, one currently needs to put documentation at various places depending on if it's a function or a type constructor.

It should be noted that back in May, it was possible to use ||| everywhere (when :doc was not yet implemented), so something in the parser made it more restrictive since then (yes, I was using ||| even when :doc was not a thing! :p).

# Steps to Reproduce

```
||| Doc for functions is accepted BEFORE the public export (but ignored)
public export
||| Doc for functions is also accepted AFTER the public export (and works)
||| (Which can use several |||, all produced by :doc)
f : Nat -> Nat

||| (Doc for TypeConstructor before the public export is accepted)
public export
||| But Doc for TypeConstructor is NOT accepted after the public export
data TypeConstructor : Type where
    ||| Doc for DataConstructor is ok
    DataConstructor : TypeConstructor
```

# Expected Behavior

1) :doc f should probably produce both documentation strings given before *and* after the "public export" annotation.
2) Documentation should be allowed after the "public export" too for a type constructor, in order to be consistent with the documentation of functions

# Observed Behavior

1) :doc f ignores the first documentation string.
2) Parse error due to the documentation after the "public export" for the type constructor

Cheers,

**Comment by fabianhjr** [comment]
Hi, thanks for the issue report.

I did some work on the parser and I think it has to do with that. Will try to work on a fix this evening.

**Comment by FranckS** [comment]
Hi @fabianhjr 

Thank you for your answer. I'm glad that the report is useful. I'll look at the patch you'll produce, that will help me to dive more into the parser code.
Many thanks for your work on Idris2!

Kind regards,

Franck

**Comment by FranckS** [comment]
Just a small update with two additional information:

The same problem happens with interfaces, where ||| Doc was previously allowed after the "public export", but is now currently forbidden by the parser.
And one more thing : previously, ||| Doc was also authorized before something commented (with --) which is nice when you want to comment a constructor, or an item of an interface, without having to remove its associated doc (or to replace it by ordinary comments with --).

I must confess that I tend to write as much doc and comments than I write code and proofs!

**Comment by fabianhjr** [comment]
@gallais / @ziman, not sure if docstrings should be allowed between the visibility and the body of whatever they are for.

Since `tyDecl` can happen wherever before the definition of a function occurs (even with many tyDecls inbetween) the first example in this issue is accepted. (Would still need to connect the two docstrings, however I think this would be unecesarly complex but could be beneficial for pattern matching definitions)

Some discussion seems to be required, if you want to try a fix to this issue refer to #469

**Comment by ziman** [comment]
I think it's fine to require docstrings to be above all the modifiers like `extern` or `total` or so. The modifiers are short, while docstrings have an arbitrary length. Otherwise you can get `public` very far (even several screens) away from the type signature, which would be probably undesirable. But the docstring can never be too far away.
